subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/GroovyRecompilationSpecProvider.java

```diff
@@ -98,47 +92,51 @@ public void initializeCompilation(JavaCompileSpec spec, RecompilationSpec recomp
 
     private void prepareFilePatterns(Set<File> filesToCompile, PatternSet classesToDelete, PatternSet filesToRecompilePatterns) {
         for (File file : filesToCompile) {
-            filesToRecompilePatterns.include(relativize(file));
-
-            Collection<String> classes = sourceFileClassNameConverter.getClassNames(file);
-            for (String staleClass : classes) {
-                String path = staleClass.replaceAll("\\.", "/");
-                classesToDelete.include(path.concat(".class"));
-            }
+            sourceDirs.relativize(file).ifPresent(relativeFile -> {
+                filesToRecompilePatterns.include(relativeFile.toString());
+
+                sourceFileClassNameConverter.getClassNames(relativeFile)
+                    .stream()
+                    .map(staleClass -> staleClass.replaceAll("\\.", "/").concat(".class"))
+                    .forEach(classesToDelete::include);
+            });
         }
     }
 
-    private String relativize(File sourceFile) {
-        List<File> dirs = sourceDirs.getSourceRoots();
-        for (File sourceDir : dirs) {
-            if (sourceFile.getAbsolutePath().startsWith(sourceDir.getAbsolutePath())) {
-                return RelativePathUtil.relativePath(sourceDir, sourceFile);
-            }
-        }
-        throw new IllegalStateException("Not found " + sourceFile + " in source dirs: " + dirs);
-    }
-
     private void processOtherChanges(PreviousCompilation previous, RecompilationSpec spec) {
         if (spec.getFullRebuildCause() != null) {
             return;
         }
-        SourceFileChangeProcessor sourceFileChangeProcessor = new SourceFileChangeProcessor(previous, sourceFileClassNameConverter);
+        SourceFileChangeProcessor sourceFileChangeProcessor = new SourceFileChangeProcessor(previous);
+
         for (FileChange fileChange : sourceChanges) {
             if (spec.getFullRebuildCause() != null) {
                 return;
             }
-            spec.getFilesToCompile().add(fileChange.getFile());
-            sourceFileChangeProcessor.processChange(fileChange.getFile(), spec);
+
+            File changedFile = fileChange.getFile();
+            Optional<File> relativeFile = sourceDirs.relativize(changedFile);
```

Comment:
For getting the relative path, use `fileChange.getNormalizedPath()`. For relative path sensitivity that will return the relative path.

Reply:
OK thanks, good to know.